File: LC/763/763.py
class Solution:
    def partitionLabels(self, S):
        result = []
        start = 0

        while start < len(S):
            current = S[start]
            current_end = start
            for end in range(len(S) - 1, start, -1):
                if current == S[end]:
                    current_end = end
                    break

            print('current_end ->', current_end)
            next_start = start + 1
            seen = dict()
            seen[current] = current_end
            while next_start < current_end:
                print('next_start = ', next_start)
                next_char = S[next_start]
                if next_char == current or next_char in seen:
                    next_start += 1
                    continue

                for end in range(len(S) - 1, current_end, -1):
                    if next_char == S[end]:
                        current_end = end
                        break
                print('inner: current_end -> ', current_end)
                seen[next_char] = current_end
                next_start += 1
            print('outer: current_end -> ', current_end)
            result.append([current_end - start + 1])
            start = current_end + 1
        return result


if __name__ == "__main__":
    s = Solution()
    print(s.partitionLabels("ababcbacadefegdehijhklij"))

File: LC/763/763.java
/*
================================================================
Test cases:
#1:
"ababcbacadefegdehijhklij"
                ^
                ^
================================================================
Brute force:
================================================================
Optimal solution:
1. Initialize Map<Character, Integer> where map will save
    the character from S as a key and index as a value.
2. Iterate from i = 0 to i = S.length():
    a. Keep inputting the current character with the current index.
3. Iterate from i = 0 to i = S.length():
    a. Record the start position, int start = i
    b. From j = i to j = S.length():
        a. Move i to map.get(S.charAt(j))
        b. If i == j, break the for loop
    c. Push the ith (ndex - start + 1) to result array

Time: O(N)
Space: O(1), 26 alphabets
================================================================
*/
class Solution {
    public List<Integer> partitionLabels(String S) {
        List<Integer> result = new ArrayList<>();
        if (S == null || S.length() == 0) {
            return result;
        }
        
        // Create map
        Map<Character, Integer> map = new HashMap<>();
        
        for (int i = 0; i < S.length(); i++) {
            map.put(S.charAt(i), i);
        }
        
        int i = 0;
        while (i < S.length()) {
            int start = i;
            int j = i;
            
            i = map.get(S.charAt(i));
            
            while (j < i) {
                i = Math.max(i, map.get(S.charAt(j)));
                j++;
            }
            
            result.add(i - start + 1);
            i++;
        }
        
        return result;
    }
}
